module/input/Microphone/src/Microphone.cpp

```diff
@@ -0,0 +1,367 @@
+    assert(file_size != -1);
+    
+    char *file_buffer = (char*) malloc(file_size);
+    fread(file_buffer, file_size, 1, file);
+    fclose(file);
+    
+    char file_size_buf[64];
+    sprintf(file_size_buf, "%d\n", file_size);
+        
+    write(fd, file_size_buf, strlen(file_size_buf));
+    
+    ssize_t write_res = write(fd, file_buffer, file_size);
+    assert(write_res == file_size);
+
+    free(file_buffer);
+}
+
+
+Microphone::Microphone(std::unique_ptr<NUClear::Environment> environment) : Reactor(std::move(environment)), config{} {
+    
+    on<Configuration>("Microphone.yaml").then([this](const Configuration& cfg) {
+        // Use configuration here from file Microphone.yaml
+        this->log_level = cfg["log_level"].as<NUClear::LogLevel>();
+    });
+    
+
+    
+    if(!voice2json_transcribe_wav(&handles)) {
+        perror("failed to start voice2json");
+        return;
+    }
+    
+    on<Trigger<MicControlMsg>>().then([this](const MicControlMsg& msg) {
+        switch(msg.type) {
+            case MIC_MSG_ENABLE: 
+                enabled = true;
+            break;
+            case MIC_MSG_DISABLE:
+                enabled = false;
+            break;
+            case MIC_MSG_TEST_AUDIO:
+                char filename[1024];
+                sprintf(filename, "/home/nubots/nuhear/audio/voice_%03d.wav", msg.data);
+            
+                write_audio_to_file(handles.stdin, filename);
+            break;
+        }
+    });
+    
+    //THIS IS FOR TESTING
+    on<Trigger<SpeechIntentMessage>>().then([this](const SpeechIntentMessage &msg) {
+        //TODO needs to be freed...
+        printf("text = %s (%p)\n", msg.text.data(), msg.text.data());
+        printf("intent = %s (%p)\n", msg.intent.data(), msg.intent.data());
+        for(Slot slot : msg.slots) {
+            printf("%s = %s\n", slot.name.data(), slot.value.data());
+        }
+        //need to free the msg...
+        //delete &msg;
+    });
+
+    
+    on<IO>(handles.stdout, IO::READ).then([this] {
+        printf("fds[0].revents & POLLIN\n");
+            
+        //TODO: need to read and append to buffer until we reach a newline...
+        char buffer[0x1000];
+        ssize_t bytes_read = read(handles.stdout, buffer, sizeof(buffer)-1);
+        if(bytes_read > 0) {
+            assert((size_t)bytes_read < sizeof(buffer)); //TODO
+            buffer[bytes_read] = 0;
+                        
+            if(enabled) {            
+                char *intent_json = voice2json_recognize_intent(buffer);
+                if(!intent_json) {
+                    perror("error");
+                }
+                
+                std::unique_ptr<SpeechIntentMessage> intent = parse_voice2json_json(intent_json, strlen(intent_json));
+                emit(std::move(intent));
+
+                free(intent_json);
+            }
+            
+        }
+    });
+    
+    //TODO: have a better story on how we handle this
```

Comment:
We really don't know why the subprocess is printing something to the standard error stream. I think the most appropriate behaviour is to just print what it displays, maybe prefixed with the subprocess that's being ran. If they have a logging facility, it would probably also be a good idea to send it there instead of the standard error stream.

Reply:
we do have a logging facility. `NUClear::log(message)`. If you want, you can change the "log level" it gets logged at - as in, log it as a trace, debug, info, warning, error, or fatal message (in order of badness).

Reply:
Ok. I log the stderr message now.